### ISSUE TITLE
RZ4Intune - Detection.ps1 fails for many apps

### ISSUE THREAD

**Issue by sebawern** [rzander/ruckzuck#68]
Intune reports installation failure on some apps. So far, I've seen these apps breaking because the detection script returns exit code -1. 

- Putty
- Sysinternals Suite
- Keepass
- ConfigMgrTools

When I run the detection script for putty manually after the failed install, it returns true, and putty is installed. I think the only thing these apps have in common is the way the detection works: they all check for **HKLM:\SOFTWARE\Classes\Installer\Products**.

A lot of other apps (TeamViewer, 7zip, ...) work, they do not use this method of detection.

**Comment by sebawern** [comment]
I'll try to patch the detection.ps1 of putty to detect the app using **HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\** to confirm the detection method is the actual problem.

**Comment by sebawern** [comment]
So the issue is related to the encoding of the script detection.ps1. VS Code says it's encoded with UTF-16 LE. I tried saving the file as UTF-8 and just upload the detection script again, and it's working!

However, working detection scripts are also shown as UTF-16 LE. So I'm not sure how to troubleshoot this further. All I can say for sure is that the script is okay, but the encoding is the root cause.

**Comment by rzander** [comment]
Thanks for providing some details... Strange is that some script do work and some not...  Have you checked that there are no leading or trailing characters in the script ?

**Comment by sebawern** [comment]
Yes. I'm actually trying out my workaround with all the "broken" apps. So far, the detection can be repaired by converting detection.ps1 and uploading it again. So the problem is either caused during the creation of the script or during the upload.

**Comment by rzander** [comment]
All the scripts are stored in %TEMP%\{ContentID} on the Client your run RZ4Intune...  If you just upload a such a detection.ps1, do you still have an Issue ? or do you have to change the encoding of the script to make it work ?

**Comment by sebawern** [comment]
Yes, the issue persists if i just upload the original file again. Only after converting the files to UTF-8 the detection starts working.

**Comment by rzander** [comment]
Ok, will change the way the scripts are created and store them as UTF-8... I currently do not have any Intune managed clients around me, so it may take a few days...

**Comment by sebawern** [comment]
Thank you. If there is anything I can do to help you, please let me know!

**Comment by rzander** [comment]
I've updated the File on https://github.com/rzander/ruckzuck/releases/tag/1.7.1.2 ... It does now create UTF16 Files, but I still have no Client to test..

**Comment by sebawern** [comment]
I am going to test it tomorrow. Thank you!

**Comment by sebawern** [comment]
It seems UTF16 won't work for some files. The documentation says UTF-8 is recommended (https://docs.microsoft.com/en-us/intune/apps/apps-win32-app-management#create-assign-and-monitor-a-win32-app Step 6). VS Code still says the encoding is UTF16 LE.

Any chance you can modify RZ4Intune.exe to create UTF-8 files?

**Comment by rzander** [comment]
Tool updated in https://github.com/rzander/ruckzuck/releases/tag/1.7.1.2  to use UTF8... You can tweak the PowerShell by yourself, it's in RZ4Intune.exe.config...

**Comment by sebawern** [comment]
So I've done some testing, the updated RZIntune.exe.config results in UTF8 with BOM. I was unable to get UTF8 without BOM, because **-Encoding UTF8NoBom** is only available in PowerShell 6 or higher. 

I found some examples of .NET code to write to files using UTF8 without BOM. I am going to try this, and report back.

**Comment by rzander** [comment]
Updated https://github.com/rzander/ruckzuck/releases/tag/1.7.1.2 to generate UTF8 without BOM... I generate the Files now in C#, so the section does no longer exists in the PowerShell script in RZ4Intune.exe.config... -> Also replace RZ4Intune.exe.config !!!!

**Comment by sebawern** [comment]
The detection.ps1 looks fine now, great job! However, installation of some packages (putty for example) still fail. But there is something new to report: When I upload the original detection.ps1 again (without editing, saving, ...) and start the installation again, it's working. So, the original encoding issue is resolved. I'm still not sure why this only happens to some apps. 

I can't find any difference between broken and working apps (e.g. PuTTy (broken) and Adobe Reader DC MUI (working)).

I really appreciate all the time and effort you are putting in to resolve this issue! Thank you.

**Comment by rzander** [comment]
In "C:\ProgramData\Microsoft\IntuneManagementExtension\Logs\IntuneManagementExtension.log" I just found the Base64 encoded detection rule (of Putty). Decoded it does look like:
```PowerShell
$bRes = Test-Path 'HKLM:\SOFTWARE\Classes\Installer\Products\C2467F44E7BA86440B828E0DA8C0BBE0' if($bRes) { $true } else { $null } exit(0)
```
So it seems, the problem is that everything is on one line, the "new line" characters from the original script are gone...

**Comment by rzander** [comment]
Updated the files in https://github.com/rzander/ruckzuck/releases/tag/1.7.1.2
I've added a Semicolon ";" on every line of the Detection and Requirements script...
Putty is now working... Is there any other Software I should test ?

**Comment by rzander** [comment]
Updated https://github.com/rzander/ruckzuck/wiki/RuckZuck-for-Intune#known-issues 
It looks not so bad (I will open a new Issue for *AdobeReader DC*) as this seems to be another bug:
![image](https://user-images.githubusercontent.com/11909453/69229832-7c783f80-0b86-11ea-866f-e16e62bbfac4.png)

**Comment by sebawern** [comment]
I've tested PuTTy, ConfigMgrTools and Keepass and they all work! Thank you!

**Comment by rzander** [comment]
AdobeReader is now also fixed, so I close this Issue...
